packages/trees/backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/renderer/TreeRenderer.java

```diff
@@ -72,27 +89,75 @@ private TreeItem renderTreeItem(VariableManager treeItemVariableManager) {
         boolean deletable = this.treeDescription.getDeletableProvider().apply(treeItemVariableManager);
         boolean selectable = this.treeDescription.getSelectableProvider().apply(treeItemVariableManager);
         List<String> iconURL = this.treeDescription.getIconURLProvider().apply(treeItemVariableManager);
-        Boolean hasChildren = this.treeDescription.getHasChildrenProvider().apply(treeItemVariableManager);
 
+        if (this.loopDetected(treeItemVariableManager, id)) {
+            return this.renderWarningTreeItem(id, kind, label, iconURL, treeItemVariableManager);
+        }
+
+        Boolean hasChildren = this.treeDescription.getHasChildrenProvider().apply(treeItemVariableManager);
         List<?> children = this.treeDescription.getChildrenProvider().apply(treeItemVariableManager);
         List<TreeItem> childrenTreeItems = new ArrayList<>(children.size());
 
+        int childIndex = 0;
         boolean expanded = !children.isEmpty();
         for (Object child : children) {
             VariableManager childVariableManager = treeItemVariableManager.createChild();
+            List<Object> childAncestors = new ArrayList<Object>(treeItemVariableManager.get(ANCESTOR_IDS, List.class).orElse(List.of()));
+            childAncestors.add(id);
+            childVariableManager.put(ANCESTOR_IDS, childAncestors);
             childVariableManager.put(VariableManager.SELF, child);
+            childVariableManager.put(INDEX, childIndex++);
             childrenTreeItems.add(this.renderTreeItem(childVariableManager));
         }
-        return TreeItem.newTreeItem(id)
+        return TreeItem.newTreeItem(id).kind(kind).label(label).editable(editable).deletable(deletable).selectable(selectable).iconURL(iconURL).children(childrenTreeItems).hasChildren(hasChildren)
+                .expanded(expanded).build();
+    }
+
+    private TreeItem renderWarningTreeItem(String id, String kind, StyledString label, List<String> iconURL, VariableManager mng) {
+        LOGGER.warn("Possible infinite loop encountered during tree rendering. Item with id '{}' has already been encountered as ancestors", id);
+
+        List<StyledStringFragment> fragments = new ArrayList<>(label.styledStringFragments());
+        fragments.add(0, ITEM_RENDERING_LOOP_FRAGMENT);
+        StyledString newLabel = new StyledString(fragments);
+        List<String> newIcons = new ArrayList<>(iconURL);
+        newIcons.add("/icons/svg/warning-overlay.svg");
+
+        List<Object> idPath = mng.get(ANCESTOR_IDS, List.class).orElse(new ArrayList<>());
+        idPath.add(id);
+        String path = idPath.stream().map(Object::toString).collect(joining("::")).toString() + "#" + mng.get(INDEX, Integer.class);
+
+        // This would need to be encoded to avoid getting a problem with id containing #
+        String virtualId = "siriuscomponent://treelooprendering?itemId=" + id + "&path=" + path;
```

Comment:
Why change that much the original id?

Reply:
Same reason describe here https://github.com/eclipse-sirius/sirius-web/pull/4015#discussion_r1766391399

If I want to be "sure" to have unique id I needs to have the path of the current item.